File: abyssal/core/viewport.py
"""The ONE place where logical world units become device pixels.

RESIZE CONTRACT
---------------
A resize produces a new Viewport value and nothing else. It must not:
  * rebuild organism geometry
  * reset or re-seed simulation state
  * allocate or free any buffer, texture or surface
  * reparent, hide or destroy anything

Viewport is an immutable value object. Constructing one is a handful of
floating point operations, so it is created fresh on every single frame; there
is no cache to invalidate and therefore no cache to get wrong.

Scale is UNIFORM on both axes by construction (a single float), so a circle in
world space is a circle on screen at every window size. There is no code path
in this project that can produce a non-uniform scale.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .world import WORLD_RADIUS, WORLD_SIZE


@dataclass(frozen=True, slots=True)
class Viewport:
    """Maps world units -> widget pixels for one rectangular stage."""

    # The stage this viewport fills, in widget pixels.
    stage_x: float
    stage_y: float
    stage_w: float
    stage_h: float

    # Uniform world-units -> pixels factor, and the pixel centre of the stage.
    scale: float
    cx: float
    cy: float

    @staticmethod
    def for_stage(x: float, y: float, w: float, h: float) -> "Viewport":
        """Fit the WORLD_SIZE square into the stage rect, letterboxed, centred.

        Uses min(w, h) so the world square is never stretched. The organism is
        centred in the stage, which keeps all four lobes symmetric about the
        stage centre regardless of the stage's aspect ratio.
        """
        w = max(float(w), 1.0)
        h = max(float(h), 1.0)
        scale = min(w, h) / WORLD_SIZE
        return Viewport(
            stage_x=float(x),
            stage_y=float(y),
            stage_w=w,
            stage_h=h,
            scale=scale,
            cx=float(x) + w / 2.0,
            cy=float(y) + h / 2.0,
        )

    # -- conversions ------------------------------------------------------
    def px(self, wx: float, wy: float) -> tuple[float, float]:
        """World point -> widget pixel point."""
        return (self.cx + wx * self.scale, self.cy + wy * self.scale)

    def length(self, world_len: float) -> float:
        """World length -> pixel length (same on both axes, by contract)."""
        return world_len * self.scale

    # -- diagnostics ------------------------------------------------------
    @property
    def organism_px_radius(self) -> float:
        return WORLD_RADIUS * self.scale

    @property
    def clips(self) -> bool:
        """True if the organism's design radius cannot fit in the stage."""
        r = self.organism_px_radius
        return (r > self.stage_w / 2.0 + 0.5) or (r > self.stage_h / 2.0 + 0.5)

    def describe(self) -> str:
        return (
            f"stage {self.stage_w:.0f}x{self.stage_h:.0f}"
            f" @{self.stage_x:.0f},{self.stage_y:.0f}"
            f"  scale {self.scale:.4f}"
            f"  r {self.organism_px_radius:.0f}px"
        )


def isotropy_error(vp: Viewport) -> float:
    """GATE 1 helper: max deviation from circularity of a unit world circle.

    Returns 0.0 for a correct viewport. Any non-zero value is a bug.
    """
    pts = [vp.px(math.cos(t) * 100.0, math.sin(t) * 100.0) for t in
           [i * math.tau / 64 for i in range(64)]]
    radii = [math.hypot(px - vp.cx, py - vp.cy) for px, py in pts]
    return max(radii) - min(radii)

File: rust/src/viewport.rs
//! The ONE place where logical world units become device pixels.
//! Port of `core/viewport.py`.
//!
//! RESIZE CONTRACT: a resize produces a new Viewport value and nothing else.
//! Scale is UNIFORM on both axes by construction (a single float), so a circle
//! in world space is a circle on screen at every window size.

use crate::world::{WORLD_RADIUS, WORLD_SIZE};

/// Maps world units -> widget pixels for one rectangular stage.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Viewport {
    /// The stage this viewport fills, in widget pixels.
    pub stage_x: f64,
    pub stage_y: f64,
    pub stage_w: f64,
    pub stage_h: f64,

    /// Uniform world-units -> pixels factor, and the pixel centre of the stage.
    pub scale: f64,
    pub cx: f64,
    pub cy: f64,
}

impl Viewport {
    /// Fit the WORLD_SIZE square into the stage rect, letterboxed, centred.
    pub fn for_stage(x: f64, y: f64, w: f64, h: f64) -> Viewport {
        let w = w.max(1.0);
        let h = h.max(1.0);
        let scale = w.min(h) / WORLD_SIZE;
        Viewport {
            stage_x: x,
            stage_y: y,
            stage_w: w,
            stage_h: h,
            scale,
            cx: x + w / 2.0,
            cy: y + h / 2.0,
        }
    }

    /// World point -> widget pixel point.
    #[inline]
    pub fn px(&self, wx: f64, wy: f64) -> (f64, f64) {
        (self.cx + wx * self.scale, self.cy + wy * self.scale)
    }

    /// World length -> pixel length (same on both axes, by contract).
    #[inline]
    pub fn length(&self, world_len: f64) -> f64 {
        world_len * self.scale
    }

    pub fn organism_px_radius(&self) -> f64 {
        WORLD_RADIUS * self.scale
    }

    /// True if the organism's design radius cannot fit in the stage.
    pub fn clips(&self) -> bool {
        let r = self.organism_px_radius();
        r > self.stage_w / 2.0 + 0.5 || r > self.stage_h / 2.0 + 0.5
    }
}

/// GATE 1 helper: max deviation from circularity of a unit world circle.
/// Returns 0.0 for a correct viewport. Any non-zero value is a bug.
pub fn isotropy_error(vp: &Viewport) -> f64 {
    let mut radii = [0.0f64; 64];
    for (i, r) in radii.iter_mut().enumerate() {
        let t = i as f64 * std::f64::consts::TAU / 64.0;
        let (px, py) = vp.px(t.cos() * 100.0, t.sin() * 100.0);
        *r = (px - vp.cx).hypot(py - vp.cy);
    }
    let mn = radii.iter().cloned().fold(f64::INFINITY, f64::min);
    let mx = radii.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    mx - mn
}
